run/run_protein_mpnn.py

```diff
@@ -179,6 +213,38 @@ def run_protein_mpnn(args):
                             score_print = np.format_float_positional(np.float32(score), unique=False, precision=4)
                             seq_rec_print = np.format_float_positional(np.float32(seq_recovery_rate.detach().cpu().numpy()), unique=False, precision=4)
                             f.write('>T={}, sample={}, score={}, seq_recovery={}\n{}\n'.format(temp,b_ix,score_print,seq_rec_print,seq)) #write generated sequence
+
+                            # FOR DESIGNED SEQUENCES
+                            if 'chain_key' in design_specs_dict:
+                                # split generated seq and save each to separate file
+                                for fsplit, af2split in zip(split_ali, split_af2):
+                                    with open(fsplit, 'a') as fsp:
+                                        fbase = os.path.splitext(os.path.basename(fsplit))[0]
+                                        chains_original = list(design_specs_dict['chain_key'][fbase].keys())
+                                        chains_to_recover = list(design_specs_dict['chain_key'][fbase].values())
+                                        chain_idx = [masked_list.index(ctr) for ctr in chains_to_recover]
+                                        seq_split = '/'.join([seq.split('/')[idx] for idx in chain_idx])
+
+                                        scores_per_chain, seq_recovery_per_chain = {}, {}
+                                        for ctr, idx in zip(chains_original, chain_idx):
+                                                # grab the indices for scoring each chain
+                                                score_idx = chain_encoding_all == idx + 1
+                                                chain_score = torch.sum(scores_per_res[score_idx], dim=-1) / torch.sum(mask_for_loss[score_idx], dim=-1)
+                                                chain_score = np.format_float_positional(np.float32(chain_score.cpu().data.numpy().mean()), unique=False, precision=4)
+                                                scores_per_chain[ctr] = chain_score
+                                                # use this for seq recovery too
+                                                chain_seq_rec = torch.sum(seq_recovery_per_res.unsqueeze(0)[score_idx], dim=-1) / torch.sum(mask_for_loss[score_idx], dim=-1)
```

Comment:
Is this `unsqueeze` in `seq_recovery_per_res.unsqueeze(0)[score_idx]` necessary? Couldn't we just do `seq_recovery_per_res`?

Reply:
Yes, the `score_idx` and `mask_for_loss` tensors have shape `[1, seq_length]`, but `seq_recovery_per_res` has shape `[seq_length],` so we get an `IndexError` if we just use `seq_recovery_per_res[score_idx]`.

So we can either add a dimension with unsqueeze:
`seq_recovery_per_res.unsqueeze(0)[score_idx]`
Or we can drop that extra dimension from `score_idx`:
`seq_recovery_per_res[torch.flatten(score_idx)]`